bpf/lib/policy.h

```diff
@@ -123,6 +123,47 @@ __policy_can_access(const void *map, struct __ctx_buff *ctx, __u32 localID,
 	}
 #endif /* ALLOW_ICMP_FRAG_NEEDED */
 
+	/* When the ICMP or ICMPv6 packet come, set its type to the key.dport*/
+	if (proto == IPPROTO_ICMP) {
+		void *data, *data_end;
+		struct iphdr *ip4;
+		__u8 icmp_type;
+		__u32 off;
+
+		if (!revalidate_data(ctx, &data, &data_end, &ip4))
+			return DROP_INVALID;
+
+		off = ((void *)ip4 - data) + ipv4_hdrlen(ip4);
+		if (ctx_load_bytes(ctx, off, &icmp_type, sizeof(icmp_type)) < 0)
+			return DROP_INVALID;
+
+		/* convert from unsigned char to unsigned short considering byteorder */
+#if __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
+		key.dport = (__u16)icmp_type;
+#else
+		key.dport = (__u16)(icmp_type << 8);
+#endif
+	} else if (proto == IPPROTO_ICMP_V6) {
+		void *data, *data_end;
+		struct ipv6hdr *ip6;
+		__u8 icmp6_type;
+		__u32 off;
+
+		if (!revalidate_data(ctx, &data, &data_end, &ip6))
+			return DROP_INVALID;
+
+		off = ((void *)ip6 - data) + ipv6_hdrlen(ctx, ETH_HLEN, &ip6->nexthdr);
+		if (ctx_load_bytes(ctx, off, &icmp6_type, sizeof(icmp6_type)) < 0)
```

Comment:
I think we can merge the ICMP and ICMPv6 blocks because the `type` field is at the same offset for both. It may also help for the complexity issue on 4.9.